### DISCUSSION TITLE
Custom relay not operating over annotated class property

### DISCUSSION THREAD

**Discussion by addisonbeck** [AutoFixture/AutoFixture#1262] [Q&A] [answered]
### Issue
`ISpecimenBuilder` is ignoring object properties that have the `[MaxLength]` attribute

### Description
I'm trying to implement a version of the [`StringLengthAttributeRelay()`](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/DataAnnotations/StringLengthAttributeRelay.cs) that looks for `[MaxLength]` instead of `[StringLength]`. The below `ISpecimenBuilder` and `TestObject()` does not work as expected. The `ISpecimenBuilder` does not ever get called for `TestObject().ThisPropertyIsSkipped`. The other two strings on the class that don't have `[MaxLength]` get run through the relay as expected.

**Class**
```
public class TestObject
{
    public string ThisGetsPickedUp { get; set; }
    [MaxLength(100)]public string ThisPropertyIsSkipped { get; set; }
    public string ThisAlsoGetsPickedUp { get; set; }
}
```
**Relay**
```
internal class MaxLengthStringRelay: ISpecimenBuilder
{
    public object Create(object request, ISpecimenContext context)
    {
        if (request == null)
        {
            return new NoSpecimen();
        }

        if (context == null)
        {
            throw new ArgumentNullException(nameof(context));
        }

        var p = request as PropertyInfo;
        if (p == null)
        {
            return new NoSpecimen();
        }

        var maxLengthAttribute = p
            .GetCustomAttributes(typeof(MaxLengthAttribute), inherit: true)
            .Cast<MaxLengthAttribute>()
            .SingleOrDefault();

        if (maxLengthAttribute == null)
        {
            return new NoSpecimen();
        }

        return context.Resolve(new ConstrainedStringRequest(maxLengthAttribute.Length));
    }
}
```
**Test**
```
[Fact]
public void TestStringLengths()
{
    var fixture = new Fixture();
    fixture.Customizations.Add(new MaxLengthStringRelay());
    var obj = fixture.Create<TestObject>();
    Assert.True(obj.ThisPropertyIsSkipped.Length == 100);
}
```

**Comment by addisonbeck** [comment]
Okay, I was able to get the relay to pick up by inserting it above the other DataAnnotation relays provided by Autofixture. In doing that I saw that there is already a relay for MaxLength as part of the library. That said, the MinAndMaxLengthAttributeRelay doesn't seem to be working for me with the above code sans custom relay.

**Comment by aivascu** [comment, marked as answer]
@addisonbeck your sepcification builder is not working since the request is being handled by the `MinAndMaxLengthAttributeRelay` which is working fine. It is not working as you expect though. 
What the `MaxLengthAttribute` does it sets an upper bound on the value. It does not enforce length. If you want a string with a minimum length of 100 you shoud specify also the `MinLengthAttribute`.

```cs
[Fact]
public void TestStringLengths()
{
    var fixture = new Fixture();

    var obj = fixture.Create<TestObject>();

    Assert.True(obj.ThisPropertyIsSkipped.Length > 100 && obj.ThisPropertyIsSkipped.Length < 200);
}
public class TestObject
{
    public string ThisGetsPickedUp { get; set; }

    [MinLength(100)]
    [MaxLength(200)]
    public string ThisPropertyIsSkipped { get; set; }

    public string ThisAlsoGetsPickedUp { get; set; }
}
```

**Comment by addisonbeck** [comment]
Hey, nice, I was just about to say setting MinLength to 100 seems to make the default relay do what I'm expecting. Thanks!
